### PR TITLE
LOK-2229 : Add  Alerts By Node support

### PR DESCRIPTION
**Pull request by MusaidAli**
## Jira link(s)
https://opennms.atlassian.net/browse/LOK-2229

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: alert/src/main/java/org/opennms/horizon/alertservice/db/entity/Alert.java [hunk 1]
```diff
@@ -40,6 +40,10 @@
 import org.opennms.horizon.alerts.proto.ManagedObjectType;
 import org.opennms.horizon.alerts.proto.Severity;
 
+import jakarta.persistence.JoinColumn;
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/db/entity/Alert.java:43]: remove unused import

File: alert/src/main/java/org/opennms/horizon/alertservice/db/entity/Alert.java [hunk 2]
```diff
@@ -138,4 +143,7 @@ public class Alert implements Serializable {
     public void incrementCount() {
         counter++;
     }
+
+    @Column(name = "node_id")
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/db/entity/Alert.java:147]: we should add this to alarm table with liquibase, see resources

File: alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java [hunk 3]
```diff
@@ -340,6 +338,30 @@ public void alertCounts(com.google.protobuf.Empty request,
...
 
+    @Override
+    public void getRecentAlertsByNode(AlertRequestByNode request, StreamObserver<ListAlertsResponse> responseObserver) {
+       try {
+
+
+           String tenantId = tenantLookup.lookupTenantId(Context.current()).orElseThrow();
+           List<org.opennms.horizon.alertservice.db.entity.Alert> alerts = alertRepository.findByNodeIdAndTenantId(request.getNodeId(), tenantId);
+
+           List<Alert> protoList = getEnrichAlertProto2nd(alerts);
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java:349]: you can simplify this with Streams: 
alerts.stream()
                      .map(alert -> Alert.newBuilder(alertMapper.toProto(alert)).build())
                      .collect(Collectors.toList());

or update method name `maptoAlertProtoList`

File: alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEventProcessor.java [hunk 4]
```diff
@@ -312,6 +312,13 @@ private org.opennms.horizon.alertservice.db.entity.Alert createNewAlert(Event ev
         } else {
             alert.setManagedObjectType(ManagedObjectType.UNDEFINED);
         }
+        if (event.getNodeId() > 0) {
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEventProcessor.java:315]: duplicate code, just need to add `alert.setNodeId(event.getNodeId());`

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java [hunk 5]
```diff
@@ -169,4 +169,19 @@ public Mono<Long> countAlertByRuleId(Long id, @GraphQLEnvironment ResolutionEnvi
     public Mono<AlertCount> getAlertCounts(@GraphQLEnvironment ResolutionEnvironment env) {
          return Mono.just(alertsClient.countAlerts(headerUtil.getAuthHeader(env)));
     }
+    @SuppressWarnings("squid:S107")
+    @GraphQLQuery(name = "getRecentAlertsByNode")
+    public Mono<ListAlertResponse> getRecentAlertsByNode(@GraphQLArgument(name = "nodeId") long nodeId,
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java:174]: let's take care of pagination in another PR

File: shared-lib/alert/src/main/proto/alerts.proto [hunk 6]
```diff
@@ -227,6 +235,7 @@ service AlertService {
   rpc deleteAlert(AlertRequest) returns(DeleteAlertResponse) {};
   rpc countAlerts(ListAlertsRequest) returns (CountAlertResponse) {};
   rpc alertCounts(google.protobuf.Empty) returns (AlertCount) {};
+  rpc getRecentAlertsByNode(AlertRequestByNode)  returns (ListAlertsResponse) {};
```
Comment (cgorantla) [inline shared-lib/alert/src/main/proto/alerts.proto:238]: we can probably omit `Recent` in the name

**Review by cgorantla — CHANGES_REQUESTED** [review]
Please add a test case in Alerts component.

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java [hunk 7]
```diff
@@ -169,4 +169,21 @@ public Mono<Long> countAlertByRuleId(Long id, @GraphQLEnvironment ResolutionEnvi
     public Mono<AlertCount> getAlertCounts(@GraphQLEnvironment ResolutionEnvironment env) {
          return Mono.just(alertsClient.countAlerts(headerUtil.getAuthHeader(env)));
     }
+    @SuppressWarnings("squid:S107")
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java:172]: do we need this `@SuppressWarnings("squid:S107")`

File: alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java [hunk 8]
```diff
@@ -340,6 +341,52 @@ public void alertCounts(com.google.protobuf.Empty request,
...
+
+        // Create a PageRequest object based on the page size, next page, filter, and sort parameters
+        Sort.Direction sortDirection = sortAscending ? Sort.Direction.ASC : Sort.Direction.DESC;
+        Pageable pageRequest = PageRequest.of(page, pageSize, Sort.by(sortDirection, sortBy));
+
+        // Get Filters
+        List<Date> timeRange = new ArrayList<>();
+        List<Severity> severities = new ArrayList<>();
+        List<String> filterNodeIds = new ArrayList<>();
+        getFilterOptions(request, timeRange, severities, filterNodeIds);
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java:362]: i think there is no requirement for filters here.  we can remove that

**Review by cgorantla — COMMENTED** [review]
Please add a test case in alert.feature

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java (hunk 7 above)
Comment (MusaidAli) [reply rest-server/src/main/java/org/opennms/horizon/server/service/GrpcAlertService.java:172]: ok ,i will remove it.

File: alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEventProcessor.java (hunk 4 above)
Comment (MusaidAli) [reply alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEventProcessor.java:316]: hi , this code was already there if(even.getNodeId()>0)

File: alert/src/main/resources/db/changelog/hs-0.1.0/tables/alert.xml [hunk 9]
```diff
@@ -59,6 +59,8 @@
             <column name="managed_object_type" type="text"/>
 
             <column name="managed_object_instance" type="text"/>
+            <column name="node_id" type="bigint"/>
```
Comment (cgorantla) [inline alert/src/main/resources/db/changelog/hs-0.1.0/tables/alert.xml:62]: we can't directly add  node_id in the table, please add new changeSet

File: alert/src/test/java/org/opennms/horizon/alertservice/stepdefs/AlertTestSteps.java [hunk 10]
```diff
@@ -270,6 +270,18 @@ public void countAlertsForTenantFilteredBySeverity(String severity, int expected
...
+    @Then("List alerts for the node")
+    public void listAlertsForNode() throws InterruptedException {
+        var requestBuilder =  AlertRequestByNode.newBuilder()
+            .setSortBy("id")
+            .setNodeId(1L)
+            .setPageSize(10)
+            .setPage(0)
+            .setSortAscending(true);
+        clientUtils.getAlertServiceStub()
+            .getAlertsByNode(requestBuilder.build());
```
Comment (cgorantla) [inline alert/src/test/java/org/opennms/horizon/alertservice/stepdefs/AlertTestSteps.java:282]: what are you validating here ?

**Pull request closed without merging**